### PR TITLE
[4.0] Menu items expiration

### PR DESCRIPTION
**Pull request by alikon**
rebase of  #17709 on 4.0

### Summary of Changes

add publish_up, publish_down for (site) menu items

### Testing Instructions
see #17709

### Documentation Changes Required
?

### CODE REVIEW

File: administrator/components/com_menus/Table/MenuTable.php [hunk 1]
```diff
@@ -47,4 +48,51 @@ public function delete($pk = null, $children = false)
...
+
+				return false;
+			}
+
+			if ((int) $this->home)
+			{
+				// Set the publish down/up always for home.
+				$this->publish_up   = $db->getNullDate();
+				$this->publish_down = $db->getNullDate();
+			}
```
Comment (Quy) [inline administrator/components/com_menus/Table/MenuTable.php]:
```suggestion
			}
```

File: administrator/components/com_menus/forms/item.xml [hunk 2]
```diff
@@ -77,6 +77,34 @@
 			<option value="-2">JTRASHED</option>
 		</field>
 
+		<field
+			name="publish_up"
+			type="calendar"
+			label="JGLOBAL_FIELD_PUBLISH_UP_LABEL"
+			description="JGLOBAL_FIELD_PUBLISH_UP_DESC"
```
Comment (Quy) [inline administrator/components/com_menus/forms/item.xml]:
```suggestion
```

File: administrator/components/com_menus/forms/item.xml [hunk 3]
```diff
@@ -77,6 +77,34 @@
...
+			size="22"
+			filter="user_utc"
+			showon="home:0"
+		/>
+
+		<field
+			name="publish_down"
+			type="calendar"
+			label="JGLOBAL_FIELD_PUBLISH_DOWN_LABEL"
+			description="JGLOBAL_FIELD_PUBLISH_DOWN_DESC"
```
Comment (Quy) [inline administrator/components/com_menus/forms/item.xml]:
```suggestion
```

**Comment by uglyeoin** [comment]
Happy to test whenever this is ready thanks for all your hard work

**Comment by Quy** [comment]
@uglyeoin It is ready to be tested.

**Comment by uglyeoin** [comment]
> @uglyeoin It is ready to be tested.

@Quy Awesome, I'll test next week when I have some time.  It says it's out of date with the base, should I wait until that is fixed?

**Comment by infograf768** [comment]
Updating branch

**Comment by uglyeoin** [comment]
It's out of date again.  Should I test anyway @infograf768 or would it mean I need to test again so I should wait?

**Comment by uglyeoin** [comment]
I have tested this item :red_circle: unsuccessfully on 3912488a53747311482445c2384623869f1ef03f<br /><br />I used Joomla! 4 and I used Patch Tester to install the patch (which might be the wrong thing to do).

I got this error when I clicked on **menus > show all menu items**

An error has occurred.
500 Unknown column 'a.publish_up' in 'field list'<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24416">issues.joomla.org/tracker/joomla-cms/24416</a>.</sub>

**Comment by wojsmol** [comment]
@uglyeoin Branch is out-of-datu but don't have merge conflicts so you can test this.

**Comment by uglyeoin** [comment]
@wojsmol in which case... it failed the test.  It installed the patch correctly so I don't think we can blame patch tester but I really am not experienced enough to know if that's right.  It does seem like the error comes from this pull request, hopefully it's something simple.

**Comment by wojsmol** [comment]
@uglyeoin You must fix database after installing the patch using the Patch Tester.

File: libraries/src/Menu/SiteMenu.php [hunk 4]
```diff
@@ -80,6 +80,8 @@ public function load()
 	{
 		$loader = function ()
 		{
+			$nulldate    = $this->db->quote($this->db->getNullDate());
+			$currentDate = Factory::getDate()->toSql();
```
Comment (Quy) [inline libraries/src/Menu/SiteMenu.php:84]: How about adding `$this->db->quote()` here and removing it from below?

Comment (alikon) [reply libraries/src/Menu/SiteMenu.php:84]: all the query should be "quoted", i'll do it if this pr will land in core

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on add416283f8637c655b6d62d4ab44308e668d3df<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24416">issues.joomla.org/tracker/joomla-cms/24416</a>.</sub>

**Comment by uglyeoin** [comment]
image
Probably not the fault of this pull request but tooltip has <br> which is ignored and shown as literal

**Comment by Quy** [comment]
@uglyeoin It is a separate issue reported in #21373.

**Comment by uglyeoin** [comment]
@Quy I don't want to clog up this pull request but I found some other things I'd like to report/suggest.  One is for the J4 Template team.  Is there some way I can easily find out where to report this stuff?  Sorry to ask such a noob question.

**Comment by Quy** [comment]
Report to the backend template repo: https://github.com/joomla/backend-template/tree/release

**Comment by infograf768** [comment]
We still need another tester here if we want to get this feature in.

**Comment by uglyeoin** [comment]
I have tested this item :white_check_mark: successfully on e665b113adaf7336255d6bd81b37d7745a28c158<br /><br />The item was shown with the correct icon.  After the publish date passed it showed as published on both the front and back end.

After the publish end date had passed it showed as unpublished on both the back and front end.

Everything worked as I expected it to.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24416">issues.joomla.org/tracker/joomla-cms/24416</a>.</sub>

**Comment by uglyeoin** [comment]
> We still need another tester here if we want to get this feature in.

done

**Comment by infograf768** [comment]
@alikon 
Shall we RTC this now or you want to quote db stuff first?

**Comment by alikon** [comment]
it has been around since 3.8.0,  so please RTC.....
i re-swear  i'll do the quote stuff after land in 4.0  :stuck_out_tongue:

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24416">issues.joomla.org/tracker/joomla-cms/24416</a>.</sub>

**Comment by uglyeoin** [comment]
Yay, thank you @alikon this is an awesome new feature

**Comment by wilsonge** [comment]
Thanks. This is a nice feature!

**Pull request merged**